### PR TITLE
NFC: Improved MFC emulation on some readers

### PR DESCRIPTION
**Pull request by AloneLiberty**
# What's new

- HALT don't disable field (because in the real world the card cannot turn itself off)

- Fixes for faster re-select of card (because we know that in these cases the reader will read again)

- Removed not particularly necessary warning output, in general improves the emulation (`hf mf fchk`, Android with different keys)

- Add trace log if reader send RATS

- Added check if the block is read, if not, it won't access it (won't reply zeros)

- Added check if we access a block we're authorized (don't allow access outside current sector)

# Verification 

- `nfc-mfclassic r a u dump` on ACR122U should work now

- Emulation should work on more Android devices (at least now works on mine)

- `hf mf chk`, `hf mf fchk` and even `hf mf hardnested` proxmark3 commands works stable

- Test for regression on other readers

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by giacomoferretti** [comment]
Works for me!
Tested on ACR122U, gonna test it soon with MCT on Android.

Should fix #1862 and maybe even #2577.

File: lib/nfc/protocols/mifare_classic.c [hunk 1]
```diff
@@ -882,12 +882,17 @@ bool mf_classic_emulator(
             break;
         }
 
-        if(cmd == 0x50 && plain_data[1] == 0x00) {
+        if(cmd == NFCA_CMD_HALT && plain_data[1] == 0x00) {
             FURI_LOG_T(TAG, "Halt received");
-            furi_hal_nfc_listen_sleep();
-            command_processed = true;
+            return false;
```
Comment (gornekich) [inline lib/nfc/protocols/mifare_classic.c]: Please, avoid return from the body of function. It's not safe if you allocate memory above return and exit from function before freeing that memory. Also it make code less readable.

I suggest rename variable `command_processed` -> `send_nack` and add new variable `need_reset`. Rework every place where you write `return` with filling these variables then call `break`. And return `need_reset` at the and of the function

File: lib/nfc/protocols/mifare_classic.c [hunk 2]
```diff
@@ -902,8 +907,7 @@ bool mf_classic_emulator(
                     access_key = MfClassicKeyA;
                 } else {
                     FURI_LOG_D(TAG, "Key not known");
-                    command_processed = true;
-                    break;
+                    return false;
```
Comment (gornekich) [inline lib/nfc/protocols/mifare_classic.c]: Same here and below

**Review by gornekich — CHANGES_REQUESTED** [review]
Hello @AloneLiberty ! Thanks for PR. Tested it and it works good.

Please, make some changes I mentioned and we will merge your PR

**Review by gornekich — APPROVED** [review]
Thanks

**Review by skotopes — APPROVED** [review]

**Pull request merged**